### ISSUE TITLE
Feature Request: Gradient for QR op

### ISSUE THREAD

**Issue by kstant0725** [tensorflow/tensorflow#6504]
The QR op is currently implemented but has no gradient.  It would be very useful if the gradient were defined so that the op could be used in networks and cost functions.

**Comment by rmlarsen** [comment]
I am not aware of anybody working on this ATM. Care to send us a PR @kstant0725 ?

**Comment by krishpop** [comment]
@kstant0725 Are there any cost functions/models that would benefit from QR decomposition? How else could this be used in training a network?

**Comment by kstant0725** [comment]
@rmlarsen and @krishpop The need for a QR decomposition came up my research when we wanted to implement a network layer that acts to orthonormalize the layer below it.  This can be a useful constraint for  training networks to find eigenfunctions.  SVD could also accomplish this, but has a higher computational cost, and imposes an additional rotational constraint.

**Comment by Queequeg92** [comment]
@kstant0725 Have you found a gradient implement for QR decomposition?

**Comment by Queequeg92** [comment]
I have checked the implement detail of QR decomposition in tensorflow. The QR decomposition in tensorflow simply calls the QR function in Eigen. But unfortunately there is no gradient implement for QR decomposition in Eigen, which indirectly results in the absence of gradient for QR decomposition in tensorflow. The QR decomposition is based on some basic matrix operations. Thus it's theoretically differentiable. The typical QR decomposition methods like Givens Rotations and Householder reflections include an iterative process, which makes the gradient very complex. There is no automatic gradient mechanism in Eigen. I think that's why there is no gradient implement for QR decomposition in Eigen. Based on the above observation, it's better to implement QR decomposition algorithm using available tensorflow operations and make use of the automatic gradient mechanism in tensorflow to obtain the gradient for QR decomposition. What do you think? @kstant0725 @rmlarsen

**Comment by kstant0725** [comment]
So far I have been using the following to orthogonalize:

    AA =tf.matmul(tf.transpose(A),A)
    L = tf.cholesky(AA)
    U = K.transpose(tf.matrix_solve(L,tf.transpose(A)))

This works well most of the time unless the matrix A is ill conditioned or not full rank.  Do you have any suggestions for an alternative which will wont throw an exception when the matrix is not full rank?

**Comment by guivenca** [comment]
I tried a simple implementation of  the gradient expression based on an article and it seemed to match the numerical gradients for the QR decomposition of tall and square random matrices. For fat ones (width>height) the results weren't as good, which might be associated with some assumption in the reference, numerical difficulties (since it requires the computation of a pseudo inverse instead of an exact one) or a mistake in the code.

Any suggestions?

```python
from tensorflow.python.framework import ops
@ops.RegisterGradient("Qr")
def _qr_grad(op, Qb,Rb):
  """Computes and returns the gradient of a QR decomposition

  computes the expression (in Matlab terms)
  Q*(Rb+( Pl.*(R*Rb'-Rb*R'+Q'*Qb-Qb'*Q))pinv(R)')
  where Pl is 1 below the diagonal and 0 elsewhere

  reference in http://dx.doi.org/10.1080/10556788.2011.610454 eq (13)

  Args:
    op: operation
    Qb: gradient in Q
    Rb: gradient in R

  Returns:
    Ab: gradient in Q*R
  """

  Q = op.outputs[0]
  R = op.outputs[1]
  batch_shape = Q.shape[0:-2]
  tr_ord = list(range(len(Q.shape)))
  tr_ord[-1],tr_ord[-2] = tr_ord[-2],tr_ord[-1]

  M1 =    tf.matmul(R,tf.transpose(Rb,tr_ord))\
        - tf.matmul(Rb,tf.transpose(R,tr_ord))\
        + tf.matmul(tf.transpose(Q,tr_ord),Qb)\
        - tf.matmul(tf.transpose(Qb,tr_ord),Q)

  # lower triangular part without the diagonal
  # equivalent to Haddamard product with Pl
  M1_low = tf.matrix_band_part(M1,-1,0)-tf.matrix_band_part(M1,0,0)

  # equivalent to explicitly computing the pseudo inverse
  # ideally we should use the fact that R is triangular
  M2_tr = tf.matrix_solve_ls(r,tf.transpose(M1_low,tr_ord),fast=False)


  M2 = tf.transpose(M2_tr,tr_ord)
  Ab = tf.matmul(Q,Rb+M2)
  return Ab
```

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by davidruegamer** [comment]
@D-Roberts Many thanks for you implementation :pray: Works like a charm for me. Any chance that there will also be a gradient implementation for `nrows > ncols` and `full_matrices = True`? Currently looking for a TF option to get the basis of the null space of a deep matrix, but the usual options (Q_2 in [Q_1 Q_2] [R_1^T R_2^T]^T in the QR decomposition or U in U Sigma V^T in the SVD decomposition) are not supported.

**Comment by D-Roberts** [comment]
@davidruegamer Glad to hear this is useful. As much as I enjoy developing tf open source, I don't have current plans for the full_matrices case. However, there are some theoretical considerations (unpublished yet) around it that I am happy to share if you'd like to email me at denisa.roberts@denisaroberts.me .

**Comment by spapanic** [comment]
> Any chance that there will also be a gradient implementation for `nrows > ncols` and `full_matrices = True` soon?

I have been working on the theoretical aspects of the tall & full case (see https://arxiv.org/abs/2409.13374), this could be used towards an implementation.
